From 03cd3a82b39a43660f4b4f37738b7f4006c657d8 Mon Sep 17 00:00:00 2001
From: Jonas Albrecht <plonkbong100@protonmail.com>
Date: Fri, 24 Jan 2020 17:28:31 +0000
Subject: [PATCH] net: lantiq_xrx200: Add pcdu settings
X-Patchwork-Bot: notify

This patch modifie:
drivers/net/ethernet/lantiq_xrx200.c
For
Add support for for PCDU
Fix typo in tx/rx delay check.
Fix wrong port number report for fixed link.
Fix printf format string.

This is required for some Routers devices like Arcadyan VGV952CJW33-E-IR

Co-Developed-by: Holger Foerster <Hamsi2k@freenet.de>
Signed-off-by: Holger Foerster <Hamsi2k@freenet.de>
Co-Developed-by: Vitalii Kovzik <vitalii.kovzik@gmail.com>
Signed-off-by: Vitalii Kovzik <vitalii.kovzik@gmail.com>
Acked-by: Jonas Albrecht <plonkbong100@protonmail.com>
Signed-off-by: Jonas Albrecht <plonkbong100@protonmail.com>
---
 drivers/net/ethernet/lantiq_xrx200.c | 56 +++++++++++++++++++++++++++++++++++-
 1 file changed, 55 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/lantiq_xrx200.c b/drivers/net/ethernet/lantiq_xrx200.c
index f4dcdb9..d55f39e 100644
--- a/drivers/net/ethernet/lantiq_xrx200.c
+++ b/drivers/net/ethernet/lantiq_xrx200.c
@@ -133,6 +133,29 @@
 #define MII_CFG_RATE_AUTO	0x40
 #define MII_CFG_RATE_MASK	0x70
 
+/*PCDU*/
+#define MII_PCDU(p)		(MII_CFG(p) + 4)
+
+#define MII_PCDU_RXLOCK				0x8000
+#define MII_PCDU_TXLOCK				0x4000
+#define MII_PCDU_RXSEL_CLK_MASK		0x3000
+#define MII_PCDU_RXSEL_CLK_AUTO		0x0000
+#define MII_PCDU_RXSEL_CLK_RXCLK	0x1000
+#define MII_PCDU_RXSEL_CLK_CLKREF	0x2000
+#define MII_PCDU_RXINIT				0x0800
+#define MII_PCDU_RXPD				0x0400
+#define MII_PCDU_RXDLY_MASK			0x0380
+
+#define MII_PCDU_TXSEL_CLK_MASK		0x0060
+#define MII_PCDU_TXSEL_CLK_AUTO		0x0000
+#define MII_PCDU_TXSEL_CLK_TXCLK	0x0020
+#define MII_PCDU_TXSEL_CLK_CLKREF	0x0040
+#define MII_PCDU_TXINIT				0x0010
+#define MII_PCDU_TXPD				0x0008
+#define MII_PCDU_TXDLY_MASK			0x0007
+
+#define MII_PCDU_DLY_MAX			0x7
+#define MII_PCDU_RXDLY_SHIFT		0x7
 /* cpu port mac */
 #define PMAC_HD_CTL		0x0000
 #define PMAC_RX_IPG		0x0024
@@ -193,6 +216,8 @@ struct xrx200_port {
 	u8 phy_addr;
 	u16 flags;
 	phy_interface_t phy_if;
+	u8 tx_delay;
+	u8 rx_delay;
 
 	int link;
 	int gpio;
@@ -1497,6 +1522,7 @@ static void xrx200_port_config(struct xrx200_priv *priv,
 		const struct xrx200_port *port)
 {
 	u16 miimode = 0;
+	bool init_delay = false;
 
 	switch (port->num) {
 	case 0: /* xMII0 */
@@ -1521,6 +1547,7 @@ static void xrx200_port_config(struct xrx200_priv *priv,
 		case PHY_INTERFACE_MODE_RGMII:
 			/* RGMII MAC mode, connected to external PHY */
 			miimode = MII_CFG_MODE_RGMII;
+			init_delay = true;
 			break;
 		default:
 			break;
@@ -1548,6 +1575,7 @@ static void xrx200_port_config(struct xrx200_priv *priv,
 		case PHY_INTERFACE_MODE_RGMII:
 			/* RGMII MAC mode, connected to external PHY */
 			miimode = MII_CFG_MODE_RGMII;
+			init_delay = true;
 			break;
 		default:
 			break;
@@ -1559,6 +1587,11 @@ static void xrx200_port_config(struct xrx200_priv *priv,
 
 	ltq_mii_w32_mask(MII_CFG_MODE_MASK, miimode | MII_CFG_EN,
 		MII_CFG(port->num));
+
+	if(init_delay)
+		ltq_mii_w32(MII_PCDU_RXLOCK | MII_PCDU_TXLOCK |
+			((port->rx_delay << MII_PCDU_RXDLY_SHIFT) & MII_PCDU_RXDLY_MASK) |
+			(port->tx_delay & MII_PCDU_TXDLY_MASK ), MII_PCDU(port->num));
 }
 
 static int xrx200_init(struct net_device *dev)
@@ -1760,6 +1793,7 @@ static void xrx200_of_port(struct xrx200_priv *priv, struct device_node *port)
 		return;
 
 	memset(p, 0, sizeof(struct xrx200_port));
+	p->num = *id;
 	p->phy_node = of_parse_phandle(port, "phy-handle", 0);
 
 	if (!p->phy_node && of_phy_is_fixed_link(port)) {
@@ -1776,9 +1810,29 @@ static void xrx200_of_port(struct xrx200_priv *priv, struct device_node *port)
 	if (!addr)
 		return;
 
-	p->num = *id;
 	p->phy_addr = *addr;
 	p->phy_if = of_get_phy_mode(port);
+	if(p->phy_if == PHY_INTERFACE_MODE_RGMII) {
+		if(!of_property_read_u8(port, "tx-delay", &p->tx_delay)){
+			if(MII_PCDU_DLY_MAX < p->tx_delay) {
+				pr_info("Invalid tx-delay value:%hhu!", p->tx_delay);
+				p->tx_delay = 1;
+			}
+		} else {
+			p->tx_delay = 1;
+		}
+		pr_info("Set tx-delay:%hhu", p->tx_delay);
+
+		if(!of_property_read_u8(port, "rx-delay", &p->rx_delay)) {
+			if( MII_PCDU_DLY_MAX < p->rx_delay){
+				pr_info("Invalid rx-delay value:%hhu!", p->rx_delay);
+				p->rx_delay = 0;
+			}
+		} else {
+			p->rx_delay = 0;
+		}
+		pr_info("Set rx-delay:%hhu", (u32)p->rx_delay);
+	}
 	if (p->phy_addr > 0x10)
 		p->flags = XRX200_PORT_TYPE_MAC;
 	else
-- 
2.7.4

